language_service/src/lib.rs

```diff
@@ -121,40 +145,77 @@ impl<'a> LanguageService<'a> {
     /// for the document, typically when the document is first opened in the editor.
     /// It should also be called whenever the source code is updated.
     ///
+    /// This is the "entry point" for the language service's logic, after its constructor.
+    ///
     /// LSP: textDocument/didOpen, textDocument/didChange
-    pub fn update_document(&mut self, uri: &str, version: u32, text: &str) {
+    pub async fn update_document(&mut self, uri: &str, version: u32, text: &str) {
         trace!("update_document: {uri} {version}");
-        let compilation = Compilation::new_open_document(
-            uri,
-            text,
+        let manifest = (self.get_manifest)(uri.to_string()).await;
```

Comment:
Have you thought about whether we might want to cache the `get_manifest` call to avoid paying the penalty of walking up the directory structure on every keystroke, even when we're not using projects? It seems like it would be a good idea, though it might require some care to make sure it doesn't get out of date. Maybe you're already doing this somewhere else and I missed it.